File: packages/core/src/discovery.ts
import { Octokit } from "@octokit/rest";
import { parse as parseYaml } from "yaml";
import { eq, sql } from "drizzle-orm";
import { repo, workflow, type Db } from "@scuttledeck/db";

/**
 * GitHub has no API to find repos using a given action, so discovery is:
 * list org repos → list .github/workflows files → fetch YAML via the
 * contents API → parse for a claude-code-action step (any owner/fork/ref).
 * Callers should reuse one EtagCache across scans so unchanged files cost
 * conditional requests only (304s don't count against the 5k/hr limit).
 */

const ACTION_REPO_NAMES = new Set(["claude-code-action", "claude-code-base-action"]);

export interface WorkflowHit {
  path: string;
  name: string | null;
  actionRef: string;
  actionVersion: string | null;
  triggers: string[];
  modelConfig: Record<string, unknown> | null;
}

export interface RepoScanResult {
  ghRepoId: number;
  fullName: string;
  defaultBranch: string | null;
  hits: WorkflowHit[];
}

/** Does this `uses:` value point at claude-code-action (any owner, any ref)? */
export function usesClaudeCodeAction(uses: string): boolean {
  const [pathPart] = uses.split("@");
  if (!pathPart) return false;
  const segments = pathPart.split("/").filter(Boolean);
  return segments.some((s) => ACTION_REPO_NAMES.has(s.toLowerCase()));
}

function extractTriggers(on: unknown): string[] {
  if (typeof on === "string") return [on];
  if (Array.isArray(on)) return on.map(String);
  if (on && typeof on === "object") return Object.keys(on);
  return [];
}

/** Parse one workflow file; returns null when it doesn't use the action. */
export function parseWorkflowFile(path: string, content: string): WorkflowHit | null {
  let doc: unknown;
  try {
    doc = parseYaml(content);
  } catch {
    return null;
  }
  if (!doc || typeof doc !== "object") return null;
  const root = doc as Record<string, unknown>;

  const jobs = root["jobs"];
  if (!jobs || typeof jobs !== "object") return null;

  for (const job of Object.values(jobs as Record<string, unknown>)) {
    if (!job || typeof job !== "object") continue;
    const steps = (job as Record<string, unknown>)["steps"];
    if (!Array.isArray(steps)) continue;
    for (const step of steps) {
      if (!step || typeof step !== "object") continue;
      const uses = (step as Record<string, unknown>)["uses"];
      if (typeof uses !== "string" || !usesClaudeCodeAction(uses)) continue;

      const at = uses.indexOf("@");
      const with_ = (step as Record<string, unknown>)["with"];
      const modelConfig: Record<string, unknown> = {};
      if (with_ && typeof with_ === "object") {
        for (const [k, v] of Object.entries(with_ as Record<string, unknown>)) {
          if (k.toLowerCase().includes("model")) modelConfig[k] = v;
        }
      }
      // `on` parses as boolean true in YAML 1.1; the yaml package (1.2) keeps it a string key.
      const on = root["on"] ?? root[String(true)];
      return {
        path,
        name: typeof root["name"] === "string" ? (root["name"] as string) : null,
        actionRef: uses,
        actionVersion: at >= 0 ? uses.slice(at + 1) : null,
        triggers: extractTriggers(on),
        modelConfig: Object.keys(modelConfig).length > 0 ? modelConfig : null,
      };
    }
  }
  return null;
}

/** Minimal GitHub surface the scanner needs — injectable for tests. */
export interface DiscoveryClient {
  listRepos(org: string): Promise<Array<{ id: number; full_name: string; default_branch: string | null }>>;
  listWorkflowPaths(owner: string, repoName: string): Promise<string[]>;
  getFileContent(owner: string, repoName: string, path: string): Promise<string | null>;
}

export class EtagCache {
  private etags = new Map<string, string>();
  private bodies = new Map<string, unknown>();
  get(key: string): { etag: string; body: unknown } | undefined {
    const etag = this.etags.get(key);
    return etag === undefined ? undefined : { etag, body: this.bodies.get(key) };
  }
  set(key: string, etag: string | undefined, body: unknown): void {
    if (etag === undefined) return;
    this.etags.set(key, etag);
    this.bodies.set(key, body);
  }
}

export function createOctokitDiscoveryClient(
  octokit: Octokit,
  cache = new EtagCache(),
): DiscoveryClient {
  async function conditionalGet<T>(key: string, fetcher: (etag?: string) => Promise<{ data: T; headers: { etag?: string } }>): Promise<T> {
    const cached = cache.get(key);
    try {
      const res = await fetcher(cached?.etag);
      cache.set(key, res.headers.etag, res.data);
      return res.data;
    } catch (err: unknown) {
      if (cached && (err as { status?: number }).status === 304) return cached.body as T;
      throw err;
    }
  }

  return {
    async listRepos(org) {
      const repos = await octokit.paginate(octokit.repos.listForOrg, {
        org,
        per_page: 100,
      });
      return repos.map((r) => ({
        id: r.id,
        full_name: r.full_name,
        default_branch: r.default_branch ?? null,
      }));
    },

    async listWorkflowPaths(owner, repoName) {
      const key = `${owner}/${repoName}:.github/workflows`;
      try {
        const data = await conditionalGet(key, (etag) =>
          octokit.repos.getContent({
            owner,
            repo: repoName,
            path: ".github/workflows",
            headers: etag ? { "if-none-match": etag } : {},
          }),
        );
        if (!Array.isArray(data)) return [];
        return data
          .filter((f) => f.type === "file" && /\.ya?ml$/.test(f.name))
          .map((f) => f.path);
      } catch (err: unknown) {
        if ((err as { status?: number }).status === 404) return [];
        throw err;
      }
    },

    async getFileContent(owner, repoName, path) {
      const key = `${owner}/${repoName}:${path}`;
      try {
        const data = await conditionalGet(key, (etag) =>
          octokit.repos.getContent({
            owner,
            repo: repoName,
            path,
            headers: etag ? { "if-none-match": etag } : {},
          }),
        );
        if (Array.isArray(data) || !("content" in data) || data.type !== "file") return null;
        return Buffer.from(data.content, "base64").toString("utf8");
      } catch (err: unknown) {
        if ((err as { status?: number }).status === 404) return null;
        throw err;
      }
    },
  };
}

export async function scanOrg(client: DiscoveryClient, org: string): Promise<RepoScanResult[]> {
  const results: RepoScanResult[] = [];
  for (const r of await client.listRepos(org)) {
    const [owner, repoName] = r.full_name.split("/") as [string, string];
    const hits: WorkflowHit[] = [];
    for (const path of await client.listWorkflowPaths(owner, repoName)) {
      const content = await client.getFileContent(owner, repoName, path);
      if (!content) continue;
      const hit = parseWorkflowFile(path, content);
      if (hit) hits.push(hit);
    }
    results.push({
      ghRepoId: r.id,
      fullName: r.full_name,
      defaultBranch: r.default_branch,
      hits,
    });
  }
  return results;
}

/** Persist a scan: upsert repos, flag has_action, upsert discovered workflows. */
export async function persistScan(
  db: Db,
  installationId: number,
  results: RepoScanResult[],
): Promise<void> {
  for (const r of results) {
    const [repoRow] = await db
      .insert(repo)
      .values({
        installationId,
        ghRepoId: r.ghRepoId,
        fullName: r.fullName,
        defaultBranch: r.defaultBranch,
        hasAction: r.hits.length > 0,
      })
      .onConflictDoUpdate({
        target: repo.ghRepoId,
        set: {
          fullName: sql`excluded.full_name`,
          defaultBranch: sql`coalesce(excluded.default_branch, ${repo.defaultBranch})`,
          hasAction: sql`excluded.has_action`,
          updatedAt: sql`now()`,
        },
      })
      .returning();

    for (const hit of r.hits) {
      await db
        .insert(workflow)
        .values({
          repoId: repoRow!.id,
          path: hit.path,
          name: hit.name,
          actionRef: hit.actionRef,
          actionVersion: hit.actionVersion,
          triggers: hit.triggers,
          modelConfig: hit.modelConfig,
          lastScannedAt: new Date(),
        })
        .onConflictDoUpdate({
          target: [workflow.repoId, workflow.path],
          set: {
            name: sql`excluded.name`,
            actionRef: sql`excluded.action_ref`,
            actionVersion: sql`excluded.action_version`,
            triggers: sql`excluded.triggers`,
            modelConfig: sql`excluded.model_config`,
            lastScannedAt: sql`now()`,
          },
        });
    }
    // Workflows that disappeared from the repo keep their rows (history), but
    // repo.has_action reflects the latest scan.
    await db
      .update(repo)
      .set({ hasAction: r.hits.length > 0 })
      .where(eq(repo.ghRepoId, r.ghRepoId));
  }
}

File: internal/discovery/discovery.go
// Package discovery finds workflows using claude-code-action across an org.
// GitHub has no API for "repos using action X", so: list repos → list
// .github/workflows files → fetch YAML via the contents API → parse for the
// action (any owner/fork/ref). ETag conditional requests keep repeat scans
// nearly free against the 5k/hr limit.
package discovery

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"gopkg.in/yaml.v3"
)

var actionRepoNames = map[string]bool{
	"claude-code-action":      true,
	"claude-code-base-action": true,
}

// UsesClaudeCodeAction reports whether a `uses:` value points at the action
// (any owner, any ref).
func UsesClaudeCodeAction(uses string) bool {
	pathPart := strings.SplitN(uses, "@", 2)[0]
	for _, seg := range strings.Split(pathPart, "/") {
		if actionRepoNames[strings.ToLower(seg)] {
			return true
		}
	}
	return false
}

type WorkflowHit struct {
	Path          string
	Name          string
	ActionRef     string
	ActionVersion string
	Triggers      []string
	ModelConfig   map[string]any
}

type RepoScanResult struct {
	GhRepoID      int64
	FullName      string
	DefaultBranch string
	Hits          []WorkflowHit
}

func extractTriggers(on any) []string {
	switch v := on.(type) {
	case string:
		return []string{v}
	case []any:
		out := make([]string, 0, len(v))
		for _, item := range v {
			out = append(out, fmt.Sprint(item))
		}
		return out
	case map[string]any:
		out := make([]string, 0, len(v))
		for k := range v {
			out = append(out, k)
		}
		return out
	}
	return nil
}

// ParseWorkflowFile returns a hit when the workflow uses the action, else nil.
func ParseWorkflowFile(path, content string) *WorkflowHit {
	var root map[string]any
	if err := yaml.Unmarshal([]byte(content), &root); err != nil {
		return nil
	}
	jobs, _ := root["jobs"].(map[string]any)
	if jobs == nil {
		return nil
	}
	for _, jobAny := range jobs {
		job, _ := jobAny.(map[string]any)
		if job == nil {
			continue
		}
		steps, _ := job["steps"].([]any)
		for _, stepAny := range steps {
			step, _ := stepAny.(map[string]any)
			if step == nil {
				continue
			}
			uses, _ := step["uses"].(string)
			if uses == "" || !UsesClaudeCodeAction(uses) {
				continue
			}

			version := ""
			if at := strings.Index(uses, "@"); at >= 0 {
				version = uses[at+1:]
			}
			modelConfig := map[string]any{}
			if with, _ := step["with"].(map[string]any); with != nil {
				for k, v := range with {
					if strings.Contains(strings.ToLower(k), "model") {
						modelConfig[k] = v
					}
				}
			}
			if len(modelConfig) == 0 {
				modelConfig = nil
			}
			// YAML 1.1 parses bare `on` as boolean true; yaml.v3 maps it to "true".
			on := root["on"]
			if on == nil {
				on = root["true"]
			}
			name, _ := root["name"].(string)
			return &WorkflowHit{
				Path:          path,
				Name:          name,
				ActionRef:     uses,
				ActionVersion: version,
				Triggers:      extractTriggers(on),
				ModelConfig:   modelConfig,
			}
		}
	}
	return nil
}

// Client is the minimal GitHub surface the scanner needs — injectable for tests.
type Client interface {
	ListRepos(ctx context.Context, org string) ([]RepoRef, error)
	ListWorkflowPaths(ctx context.Context, owner, repo string) ([]string, error)
	GetFileContent(ctx context.Context, owner, repo, path string) (string, error)
}

type RepoRef struct {
	ID            int64
	FullName      string
	DefaultBranch string
}

// ScanOrg walks every repo and returns the discovered inventory.
func ScanOrg(ctx context.Context, c Client, org string) ([]RepoScanResult, error) {
	repos, err := c.ListRepos(ctx, org)
	if err != nil {
		return nil, err
	}
	results := make([]RepoScanResult, 0, len(repos))
	for _, r := range repos {
		parts := strings.SplitN(r.FullName, "/", 2)
		owner, repoName := parts[0], parts[len(parts)-1]
		paths, err := c.ListWorkflowPaths(ctx, owner, repoName)
		if err != nil {
			return nil, err
		}
		var hits []WorkflowHit
		for _, p := range paths {
			content, err := c.GetFileContent(ctx, owner, repoName, p)
			if err != nil {
				return nil, err
			}
			if content == "" {
				continue
			}
			if hit := ParseWorkflowFile(p, content); hit != nil {
				hits = append(hits, *hit)
			}
		}
		results = append(results, RepoScanResult{
			GhRepoID:      r.ID,
			FullName:      r.FullName,
			DefaultBranch: r.DefaultBranch,
			Hits:          hits,
		})
	}
	return results, nil
}

// PersistScan upserts repos, flags has_action, and upserts workflows.
func PersistScan(ctx context.Context, pool *pgxpool.Pool, installationID int64, results []RepoScanResult) error {
	for _, r := range results {
		var repoID int64
		if err := pool.QueryRow(ctx, `
			insert into repo (installation_id, gh_repo_id, full_name, default_branch, has_action)
			values ($1, $2, $3, nullif($4, ''), $5)
			on conflict (gh_repo_id) do update set
				full_name = excluded.full_name,
				default_branch = coalesce(excluded.default_branch, repo.default_branch),
				has_action = excluded.has_action,
				updated_at = now()
			returning id`,
			installationID, r.GhRepoID, r.FullName, r.DefaultBranch, len(r.Hits) > 0).Scan(&repoID); err != nil {
			return err
		}
		for _, hit := range r.Hits {
			var modelConfig any
			if hit.ModelConfig != nil {
				b, _ := json.Marshal(hit.ModelConfig)
				modelConfig = b
			}
			triggers, _ := json.Marshal(hit.Triggers)
			if _, err := pool.Exec(ctx, `
				insert into workflow (repo_id, path, name, action_ref, action_version, triggers, model_config, last_scanned_at)
				values ($1, $2, nullif($3,''), $4, nullif($5,''), $6, $7, now())
				on conflict (repo_id, path) do update set
					name = excluded.name, action_ref = excluded.action_ref,
					action_version = excluded.action_version, triggers = excluded.triggers,
					model_config = excluded.model_config, last_scanned_at = now()`,
				repoID, hit.Path, hit.Name, hit.ActionRef, hit.ActionVersion, triggers, modelConfig); err != nil {
				return err
			}
		}
	}
	return nil
}

// restClient implements Client against api.github.com with an ETag cache.
type restClient struct {
	token  string
	http   *http.Client
	mu     sync.Mutex
	etags  map[string]string
	bodies map[string][]byte
}

func NewRESTClient(token string) Client {
	return &restClient{
		token:  token,
		http:   &http.Client{Timeout: 30 * time.Second},
		etags:  map[string]string{},
		bodies: map[string][]byte{},
	}
}

func (c *restClient) get(ctx context.Context, url string) (int, []byte, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return 0, nil, err
	}
	req.Header.Set("Accept", "application/vnd.github+json")
	if c.token != "" {
		req.Header.Set("Authorization", "Bearer "+c.token)
	}
	c.mu.Lock()
	if etag := c.etags[url]; etag != "" {
		req.Header.Set("If-None-Match", etag)
	}
	c.mu.Unlock()

	res, err := c.http.Do(req)
	if err != nil {
		return 0, nil, err
	}
	defer res.Body.Close()

	if res.StatusCode == http.StatusNotModified {
		c.mu.Lock()
		body := c.bodies[url]
		c.mu.Unlock()
		return http.StatusOK, body, nil
	}
	body, err := io.ReadAll(res.Body)
	if err != nil {
		return 0, nil, err
	}
	if res.StatusCode == http.StatusOK {
		if etag := res.Header.Get("ETag"); etag != "" {
			c.mu.Lock()
			c.etags[url], c.bodies[url] = etag, body
			c.mu.Unlock()
		}
	}
	return res.StatusCode, body, nil
}

var linkNextRe = regexp.MustCompile(`<([^>]+)>;\s*rel="next"`)

func (c *restClient) ListRepos(ctx context.Context, org string) ([]RepoRef, error) {
	var out []RepoRef
	// Works for both orgs and user accounts.
	for _, base := range []string{
		fmt.Sprintf("https://api.github.com/orgs/%s/repos?per_page=100", org),
		fmt.Sprintf("https://api.github.com/users/%s/repos?per_page=100", org),
	} {
		url := base
		for url != "" {
			status, body, err := c.get(ctx, url)
			if err != nil {
				return nil, err
			}
			if status == http.StatusNotFound {
				break // try the next base
			}
			if status != http.StatusOK {
				return nil, fmt.Errorf("GET %s: HTTP %d", url, status)
			}
			var page []struct {
				ID            int64  `json:"id"`
				FullName      string `json:"full_name"`
				DefaultBranch string `json:"default_branch"`
			}
			if err := json.Unmarshal(body, &page); err != nil {
				return nil, err
			}
			for _, r := range page {
				out = append(out, RepoRef{ID: r.ID, FullName: r.FullName, DefaultBranch: r.DefaultBranch})
			}
			url = ""
			if len(page) == 100 {
				// naive next-page: rely on Link header when present
				if m := linkNextRe.FindStringSubmatch(base); m != nil {
					url = m[1]
				}
			}
		}
		if len(out) > 0 {
			return out, nil
		}
	}
	return out, nil
}

func (c *restClient) ListWorkflowPaths(ctx context.Context, owner, repo string) ([]string, error) {
	url := fmt.Sprintf("https://api.github.com/repos/%s/%s/contents/.github/workflows", owner, repo)
	status, body, err := c.get(ctx, url)
	if err != nil {
		return nil, err
	}
	if status == http.StatusNotFound {
		return nil, nil
	}
	if status != http.StatusOK {
		return nil, fmt.Errorf("GET %s: HTTP %d", url, status)
	}
	var files []struct {
		Type string `json:"type"`
		Name string `json:"name"`
		Path string `json:"path"`
	}
	if err := json.Unmarshal(body, &files); err != nil {
		return nil, nil // single file object, not a workflows dir
	}
	var out []string
	for _, f := range files {
		if f.Type == "file" && (strings.HasSuffix(f.Name, ".yml") || strings.HasSuffix(f.Name, ".yaml")) {
			out = append(out, f.Path)
		}
	}
	return out, nil
}

func (c *restClient) GetFileContent(ctx context.Context, owner, repo, path string) (string, error) {
	url := fmt.Sprintf("https://api.github.com/repos/%s/%s/contents/%s", owner, repo, path)
	status, body, err := c.get(ctx, url)
	if err != nil {
		return "", err
	}
	if status == http.StatusNotFound {
		return "", nil
	}
	if status != http.StatusOK {
		return "", fmt.Errorf("GET %s: HTTP %d", url, status)
	}
	var file struct {
		Type    string `json:"type"`
		Content string `json:"content"`
	}
	if err := json.Unmarshal(body, &file); err != nil || file.Type != "file" {
		return "", nil
	}
	decoded, err := base64.StdEncoding.DecodeString(strings.ReplaceAll(file.Content, "\n", ""))
	if err != nil {
		return "", nil
	}
	return string(decoded), nil
}
